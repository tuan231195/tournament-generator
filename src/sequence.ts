let sequenceMap = {
	tournament: 0,
	game: 0,
	team: 0,
};

export const nextTournamentId = () => {
	return ++sequenceMap.tournament;
};

export const nextGameId = () => {
	return ++sequenceMap.game;
};

export const nextTeamId = () => {
	return ++sequenceMap.team;
};

export const reset = () => {
	sequenceMap = {
		tournament: 0,
		game: 0,
		team: 0,
	};
};
