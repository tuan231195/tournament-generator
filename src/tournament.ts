import { Game, GameData } from './game';
import { nextGameId, nextTournamentId } from './sequence';
import { Team } from './team';
import { isPowerOf2 } from './utils';
const Table = require('cli-table3');

export class Tournament {
	id!: number;
	games: Game[] = [];
	teams: Team[] = [];

	static create(teams: Team[]) {
		const tournamentId = nextTournamentId();
		const teamSize = teams.length;
		// team size is not power of 2
		if (!isPowerOf2(teamSize)) {
			throw new Error(`Invalid team size`);
		}

		let currentRoundSize = teamSize / 2;
		let roundNum = 1;

		let currentRoundGames: Game[] = [];
		let previousRoundGames: Game[] = [];

		const tournament = new Tournament();
		tournament.teams = teams;
		tournament.id = nextTournamentId();

		while (currentRoundSize >= 1) {
			for (let i = 0; i < currentRoundSize; i++) {
				const game: Game = new Game({
					type: 'win',
					tournamentId,
					id: nextGameId(),
					round: roundNum,
					...(roundNum === 1 && {
						home: teams[2 * i].id,
						away: teams[2 * i + 1].id,
					}),
				});
				tournament.games.push(game);
				currentRoundGames.push(game);
				if (roundNum > 1) {
					// when the current round number is greater than 1, we can't determine the exact participant teams. Therefore, we must save references to the previous (children) games
					const childGames = [
						previousRoundGames[2 * i],
						previousRoundGames[2 * i + 1],
					];
					game.homeFrom = childGames[0].id;
					game.awayFrom = childGames[1].id;
				}
			}
			if (currentRoundSize === 1 && previousRoundGames.length >= 2) {
				// add the 3rd place game
				const bronzeGame: Game = {
					type: 'lose',
					tournamentId,
					id: nextGameId(),
					round: roundNum,
					homeFrom: previousRoundGames[0].id,
					awayFrom: previousRoundGames[1].id,
				};
				tournament.games.push(bronzeGame);
			}
			previousRoundGames = currentRoundGames;
			currentRoundGames = [];
			currentRoundSize = currentRoundSize / 2;
			roundNum++;
		}

		return tournament;
	}

	findTeam(id: number) {
		const team = this.teams.find((team) => team.id === id);
		if (!team) {
			throw new Error(`Team ${id} not found`);
		}
		return team;
	}

	updateGame(id: number, data: GameData) {
		const game = this.games.find((game) => game.id === id);
		if (!game) {
			throw new Error(`Game with ${id} not found`);
		}
		game.court = data.court;
		game.time = data.time;
	}

	updateScore(id: number, homeScore: number, awayScore: number) {
		if (homeScore < 0 || awayScore < 0 || homeScore === awayScore) {
			throw new Error('Invalid score');
		}
		const game = this.games.find((game) => game.id === id);
		if (!game) {
			throw new Error(`Game with id ${id} not found`);
		}
		const { home, away } = game;
		if (!home || !away) {
			throw new Error('Game not settled');
		}

		game.homeScore = homeScore;
		game.awayScore = awayScore;
		const parentGames = this.games.filter(
			(parentGame) =>
				parentGame.homeFrom === game.id ||
				parentGame.awayFrom === game.id
		);
		if (!parentGames.length) {
			return;
		}
		const winParentGame = parentGames.find(
			(parentGame) => parentGame.type === 'win'
		);
		const loseParentGame = parentGames.find(
			(parentGame) => parentGame.type === 'lose'
		);
		const winningTeam = homeScore > awayScore ? home : away;
		const losingTeam = homeScore < awayScore ? home : away;

		if (winParentGame) {
			if (winParentGame.homeFrom === game.id) {
				winParentGame.home = winningTeam;
			} else {
				winParentGame.away = winningTeam;
			}
		}
		if (loseParentGame) {
			if (loseParentGame.homeFrom === game.id) {
				loseParentGame.home = losingTeam;
			} else {
				loseParentGame.away = losingTeam;
			}
		}
	}

	display(styles: Record<string, unknown> = {}) {
		console.log('-------- Current tournament result ----------');
		const table = new Table({
			head: [
				'ID',
				'Round',
				'Home',
				'Away',
				'Time',
				'Home Score',
				'Away Score',
				'Court',
			],
			...styles,
		});

		table.push(
			...this.games.map((game) => {
				let homeDescription: string;
				let awayDescription: string;
				if (game.home) {
					homeDescription = this.findTeam(game.home).name;
				} else {
					homeDescription = `${
						game.type === 'win' ? 'Winner' : 'Loser'
					} of game ${game.homeFrom}`;
				}

				if (game.away) {
					awayDescription = this.findTeam(game.away).name;
				} else {
					awayDescription = `${
						game.type === 'win' ? 'Winner' : 'Loser'
					} of game ${game.awayFrom}`;
				}
				return [
					game.id,
					game.round,
					homeDescription,
					awayDescription,
					game.time?.toISOString() ?? '',
					game.homeScore ?? '',
					game.awayScore ?? '',
					game.court ?? '',
				];
			})
		);
		console.log(table.toString());
		return table.toString();
	}
}
