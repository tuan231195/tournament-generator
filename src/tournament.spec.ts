import { reset } from './sequence';
import { Team } from './team';
import { Tournament } from './tournament';

function generateTeamData(teamSize: number) {
	return Array.from({ length: teamSize }).map((_, index) => {
		const id = index + 1;
		return {
			id: id,
			name: `Team ${id}`,
		};
	});
}

describe('Tournament', () => {
	beforeEach(() => {
		reset();
	});
	describe('Create', () => {
		let teams: Team[];

		describe('Validation', () => {
			it('should throw an error if the given team size is invalid', () => {
				const teams = generateTeamData(3);
				expect(() => Tournament.create(teams)).toThrow(
					'Invalid team size'
				);
			});
		});

		describe('More than 2 teams', () => {
			beforeEach(() => {
				teams = generateTeamData(4);
			});
			it('should create correct number of games', () => {
				const { games } = Tournament.create(teams);
				expect(games.length).toEqual(4);
			});

			it('should create correct number of games', () => {
				const { games } = Tournament.create(teams);
				expect(games).toMatchSnapshot();
			});
		});

		describe('2 teams', () => {
			beforeEach(() => {
				teams = generateTeamData(2);
			});
			it('should create correct number of games', () => {
				const { games } = Tournament.create(teams);
				expect(games.length).toEqual(1);
			});

			it('should create correct number of games', () => {
				const { games } = Tournament.create(teams);
				expect(games).toMatchSnapshot();
			});
		});
	});

	describe('Display', () => {
		let tournament: Tournament;

		describe('Complete tournament', () => {
			beforeEach(() => {
				const teams = generateTeamData(4);
				tournament = Tournament.create(teams);
				tournament.updateScore(1, 2, 1);
				tournament.updateGame(1, {
					time: new Date('2021-01-01T00:00:00Z'),
					court: 1,
				});
				tournament.updateScore(2, 3, 2);
				tournament.updateGame(2, {
					time: new Date('2021-01-02T00:00:00Z'),
					court: 2,
				});
				tournament.updateScore(3, 1, 0);
				tournament.updateGame(3, {
					time: new Date('2021-01-03T00:00:00Z'),
					court: 3,
				});
				tournament.updateScore(4, 2, 0);
				tournament.updateGame(4, {
					time: new Date('2021-01-04T00:00:00Z'),
					court: 4,
				});
			});
			it('should display tournament data correctly', () => {
				expect(
					tournament.display({
						style: {
							head: [], //disable colors in header cells
							border: [], //disable colors for the border
						},
					})
				).toMatchSnapshot();
			});
		});

		describe('Incomplete tournament', () => {
			beforeEach(() => {
				const teams = generateTeamData(4);
				tournament = Tournament.create(teams);
			});
			it('should display tournament data correctly', () => {
				expect(
					tournament.display({
						style: {
							head: [], //disable colors in header cells
							border: [], //disable colors for the border
						},
					})
				).toMatchSnapshot();
			});
		});
	});

	describe('Update score', () => {
		let tournament: Tournament;

		beforeEach(() => {
			const teams = generateTeamData(4);
			tournament = Tournament.create(teams);
		});

		describe('Validation', () => {
			it('should throw an error if the score given is invalid', () => {
				expect(() => tournament.updateScore(3, -1, 2)).toThrow(
					'Invalid score'
				);
				expect(() => tournament.updateScore(3, 1, -2)).toThrow(
					'Invalid score'
				);
				expect(() => tournament.updateScore(3, 1, 1)).toThrow(
					'Invalid score'
				);
			});

			it('should throw an error if the given game is not found', () => {
				expect(() => tournament.updateScore(7, 1, 2)).toThrow(
					'Game with id 7 not found'
				);
			});

			it('should throw an error if the home and away team are not settled', () => {
				expect(() => tournament.updateScore(3, 1, 2)).toThrow(
					'Game not settled'
				);
			});

			it('should update the score correctly', () => {
				tournament.updateScore(1, 1, 2);
				const game = tournament.games[0];
				expect(game.awayScore).toEqual(2);
				expect(game.homeScore).toEqual(1);
			});

			it.each([
				[
					{
						home: 1,
						away: 2,
						gameId: 1,
						expected: { home: 1 },
					},
				],
				[
					{
						home: 2,
						away: 1,
						gameId: 1,
						expected: { home: 2 },
					},
				],
			])(
				'should update the parent game participant correctly',
				({ home, away, gameId, expected }) => {
					tournament.updateScore(gameId, home, away);
					const game = tournament.games[3];
					expect(game).toMatchObject(expected);
				}
			);
		});
	});
});
