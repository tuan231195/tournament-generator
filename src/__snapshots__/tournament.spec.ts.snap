// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tournament Create 2 teams should create correct number of games 1`] = `
Array [
  Game {
    "away": 2,
    "home": 1,
    "id": 1,
    "round": 1,
    "tournamentId": 1,
    "type": "win",
  },
]
`;

exports[`Tournament Create More than 2 teams should create correct number of games 1`] = `
Array [
  Game {
    "away": 2,
    "home": 1,
    "id": 1,
    "round": 1,
    "tournamentId": 1,
    "type": "win",
  },
  Game {
    "away": 4,
    "home": 3,
    "id": 2,
    "round": 1,
    "tournamentId": 1,
    "type": "win",
  },
  Game {
    "awayFrom": 2,
    "homeFrom": 1,
    "id": 3,
    "round": 2,
    "tournamentId": 1,
    "type": "win",
  },
  Object {
    "awayFrom": 2,
    "homeFrom": 1,
    "id": 4,
    "round": 2,
    "tournamentId": 1,
    "type": "lose",
  },
]
`;

exports[`Tournament Display Complete tournament should display tournament data correctly 1`] = `
"┌────┬───────┬────────┬────────┬──────────────────────────┬────────────┬────────────┬───────┐
│ ID │ Round │ Home   │ Away   │ Time                     │ Home Score │ Away Score │ Court │
├────┼───────┼────────┼────────┼──────────────────────────┼────────────┼────────────┼───────┤
│ 1  │ 1     │ Team 1 │ Team 2 │ 2021-01-01T00:00:00.000Z │ 2          │ 1          │ 1     │
├────┼───────┼────────┼────────┼──────────────────────────┼────────────┼────────────┼───────┤
│ 2  │ 1     │ Team 3 │ Team 4 │ 2021-01-02T00:00:00.000Z │ 3          │ 2          │ 2     │
├────┼───────┼────────┼────────┼──────────────────────────┼────────────┼────────────┼───────┤
│ 3  │ 2     │ Team 1 │ Team 3 │ 2021-01-03T00:00:00.000Z │ 1          │ 0          │ 3     │
├────┼───────┼────────┼────────┼──────────────────────────┼────────────┼────────────┼───────┤
│ 4  │ 2     │ Team 2 │ Team 4 │ 2021-01-04T00:00:00.000Z │ 2          │ 0          │ 4     │
└────┴───────┴────────┴────────┴──────────────────────────┴────────────┴────────────┴───────┘"
`;

exports[`Tournament Display Incomplete tournament should display tournament data correctly 1`] = `
"┌────┬───────┬──────────────────┬──────────────────┬──────┬────────────┬────────────┬───────┐
│ ID │ Round │ Home             │ Away             │ Time │ Home Score │ Away Score │ Court │
├────┼───────┼──────────────────┼──────────────────┼──────┼────────────┼────────────┼───────┤
│ 1  │ 1     │ Team 1           │ Team 2           │      │            │            │       │
├────┼───────┼──────────────────┼──────────────────┼──────┼────────────┼────────────┼───────┤
│ 2  │ 1     │ Team 3           │ Team 4           │      │            │            │       │
├────┼───────┼──────────────────┼──────────────────┼──────┼────────────┼────────────┼───────┤
│ 3  │ 2     │ Winner of game 1 │ Winner of game 2 │      │            │            │       │
├────┼───────┼──────────────────┼──────────────────┼──────┼────────────┼────────────┼───────┤
│ 4  │ 2     │ Loser of game 1  │ Loser of game 2  │      │            │            │       │
└────┴───────┴──────────────────┴──────────────────┴──────┴────────────┴────────────┴───────┘"
`;
